compiler/plugins/target/AMD-AIE/aie/AMDAIEAssignBufferAddresses.cpp

```diff
@@ -46,6 +50,204 @@ static LogicalResult basicAllocation(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// BankAwareAllocation : round-robin each alloc over available banks
+//===----------------------------------------------------------------------===//
+typedef struct BankLimits {
+  int64_t startAddr;
+  int64_t endAddr;
+} BankLimits;
+
+// Function that given a number of banks and their size, computes the start and
+// end addresses for each bank and fills in the entry in the bankLimits vector.
+void fillBankLimits(int numBanks, int bankSize,
+                    std::vector<BankLimits> &bankLimits) {
+  for (int i = 0; i < numBanks; i++) {
+    int64_t startAddr = bankSize * i;
+    int64_t endAddr = bankSize * (i + 1);
+    bankLimits.push_back({startAddr, endAddr});
+  }
+}
+
+// Function that sets the address attribute of the given buffer to the given
+// start_addr. It also updates the entry in the nextAddrInBanks for the
+// corresponding bank.
+void setAndUpdateAddressInBank(BufferOp buffer, int64_t start_addr,
+                               int64_t end_addr,
+                               std::vector<int64_t> &nextAddrInBanks) {
+  buffer.setAddress(start_addr);
+  nextAddrInBanks[buffer.getMemBank().value()] = end_addr;
+}
+
+// Function that checks whether the given buffer already has a set address
+// attribute. If it does, it finds in which bank the buffer is and checks
+// whether there is enough space left for it. If there is the function returns
+// true and if not, the function emits a warning that the address will be
+// overwritten and returns false (which will cause the buffer to be added to
+// the list of buffers without addresses, to be completed later on).
+FailureOr<bool> checkAndAddBufferWithAddress(
+    BufferOp buffer, int numBanks, std::vector<int64_t> &nextAddrInBanks,
+    std::vector<BankLimits> &bankLimits) {
+  auto addrAttr = buffer->getAttrOfType<IntegerAttr>("address");
+  if (!addrAttr) return false;
+
+  int addr = addrAttr.getInt();
+  for (int i = 0; i < numBanks; i++) {
+    // If the address is not within the bank, continue.
+    if (addr < bankLimits[i].startAddr || addr >= bankLimits[i].endAddr)
```

Comment:
Nit: you can get the (first) bank that the address is on in O(1) instead of O(numBanks).

Reply:
Sorry, I didn't get this.

Reply:
You can statically determine which `i` won't be a `continue`, and so eliminate the for loop:

auto bankIndex = bankLimits[i].startAddr / bankSize;


Or you can have a function:

// Return the bank that `address` is in, where banks are non-interleaved and of size `bankSize`.
int getBankOfAddress(int address, int bankSize) {
  return address / bankSize;
}

Reply:
Oh, I see. Thanks!